transport/server.cc

```diff
@@ -207,11 +207,68 @@ cql_server::cql_server(distributed<cql3::query_processor>& qp, auth::service& au
+
+        sm::make_derive("bootstrapping_errors", _stats.errors[exceptions::exception_code::IS_BOOTSTRAPPING],
+                        sm::description("Counts the total number of raised CQL IS_BOOTSTRAPPING errors.")),
+
+        sm::make_derive("truncate_errors", _stats.errors[exceptions::exception_code::TRUNCATE_ERROR],
+                        sm::description("Counts the total number of raised CQL TRUNCATE errors.")),
+
+        sm::make_derive("write_timeouts", _stats.errors[exceptions::exception_code::WRITE_TIMEOUT],
+                        sm::description("Counts the total number of raised CQL WRITE_TIMEOUT errors.")),
+
+        sm::make_derive("read_timeouts", _stats.errors[exceptions::exception_code::READ_TIMEOUT],
+                        sm::description("Counts the total number of raised CQL READ_TIMEOUT errors.")),
+
+        sm::make_derive("read_failures", _stats.errors[exceptions::exception_code::READ_FAILURE],
+                        sm::description("Counts the total number of raised CQL READ_FAILURE errors.")),
 
+        sm::make_derive("function_failures", _stats.errors[exceptions::exception_code::FUNCTION_FAILURE],
+                        sm::description("Counts the total number of raised CQL FUNCTION_FAILURE errors.")),
+
+        sm::make_derive("write_failures", _stats.errors[exceptions::exception_code::WRITE_FAILURE],
+                        sm::description("Counts the total number of raised CQL WRITE_FAILURE errors.")),
+
+        sm::make_derive("cdc_write_failures", _stats.errors[exceptions::exception_code::CDC_WRITE_FAILURE],
+                        sm::description("Counts the total number of raised CQL CDC_WRITE_FAILURE errors.")),
+
+        sm::make_derive("syntax_errors", _stats.errors[exceptions::exception_code::SYNTAX_ERROR],
+                        sm::description("Counts the total number of raised CQL SYNTAX errors.")),
+
+        sm::make_derive("unathorized_errors", _stats.errors[exceptions::exception_code::UNAUTHORIZED],
+                        sm::description("Counts the total number of raised CQL UNATHORIZED errors.")),
+
+        sm::make_derive("invalid_errors", _stats.errors[exceptions::exception_code::INVALID],
+                        sm::description("Counts the total number of raised CQL INVALID errors.")),
+
+        sm::make_derive("config_errors", _stats.errors[exceptions::exception_code::CONFIG_ERROR],
+                        sm::description("Counts the total number of raised CQL CONFIG errors.")),
+
+        sm::make_derive("already_exists_errors", _stats.errors[exceptions::exception_code::ALREADY_EXISTS],
+                        sm::description("Counts the total number of raised CQL ALREADY_EXISTS errors.")),
+
+        sm::make_derive("unprepared_errors", _stats.errors[exceptions::exception_code::UNPREPARED],
+                        sm::description("Counts the total number of raised CQL UNPREPARED errors.")),
+        
```

Comment:
Maybe there should be just one metric, errors, with a label for the exception code.

Adding so many metrics fills me with dread.

Reply:
Adding to @avikivity and @nyh it's best to add only one metrics: errors_total (using a suffix that indicates the unit is  good practice)
with a label per error type.
Things to note:
1. It's best to have metrics initialized to zero at an early stage.
2. Labels are mutually exclusive, each error should be counted once under one of the labels, this way summing over all the labels values acts as having one global counter for all the combine errorss.
3. when you add the different metrics with their specific labels, you need to use the same description.

Reply:
Looking at the labels then